File: algorithms/0814.binary-tree-pruning/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pruneTree(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}

	if root.Left != nil {
		root.Left = pruneTree(root.Left)
	}

	if root.Right != nil {
		root.Right = pruneTree(root.Right)
	}

	if root.Val == 0 && root.Left == nil && root.Right == nil {
		return nil
	}

	return root
}

File: algorithms/0814.binary-tree-pruning/solution.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {TreeNode}
def prune_tree(root)
  return root if root.nil?

  root.left = prune_tree(root.left) if root.left
  root.right = prune_tree(root.right) if root.right

  if root.val == 0 && root.left.nil? && root.right.nil?
    return nil
  end

  root
end
